### ISSUE TITLE
Batch Norm Initialization

### ISSUE THREAD

**Issue by shubham1206agra** [mlpack/mlpack#3228]
CC: @rcurtin 

I have started the implementation of the BatchNorm layer. Please watch the files below.
https://github.com/shubham1206agra/mlpack/blob/header-only/src/mlpack/methods/ann/layer/not_adapted/batch_norm.hpp
https://github.com/shubham1206agra/mlpack/blob/header-only/src/mlpack/methods/ann/layer/not_adapted/batch_norm_impl.hpp

But now, I need to change the initialization of weights layer-wise but cannot do so due to the lack of such an API. Please suggest to me some ways implement the above.

**Comment by rcurtin** [comment]
I spent a while thinking about it.  The core of the issue is that we want to set specific initializations for the weights of the `BatchNorm` layer, *regardless* of the initialization that is being used by the network.  This means that whatever we implement must apply whenever we re-initialize the weights of the network.  Technically, any strategy that uses the `loading` parameter is incorrect: when we serialize or deserialize, this does not mean that the user wants to reinitialize the weights.  At least from my perspective, I see two basic possibilities:

#### Add as part of `SetWeights()`

One could, when `SetWeights()` is called, reset the weights of the batch normalization layer to 0 and 1.  But, the issue here is that `SetWeights()` is not guaranteed to be called every time the weights of a network are re-initialized:

https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/ffn_impl.hpp#L616-L618

The `FFN` class, for instance, only calls `SetWeights()` (via `SetLayerMemory()`) when the size of the parameters matrix is changed.  So, to make this approach work, we would need to make `FFN` (and `RNN`) call `SetWeights()` whenever parameters are re-initialized.  This can result in unnecessary computation, but it's not a huge deal, we are just making aliases of weights for every layer.  We would also want to update documentation for `SetWeights()` to indicate that this is called whenever parameters are re-initialized, so custom initializations can happen on a layer-wise basis there.

We might consider also adding a flag to `BatchNorm` so the user can choose to *not* use the default batch norm initialization.

#### Make a new function for re-initializing weights

We could add another method to `Layer` called, say, `CustomInitialize()` or something, that by default does nothing.  Then we can overload this method in `BatchNorm`.

We'd then need to ensure that from `FFN` and `RNN`, we call `CustomInitialize()` (or whatever we call the method) on each layer in the network.  (We can probably do that in `FFN::InitializeWeights()`.)  I guess `MultiLayer` will need an implementation too.  Of course that will also need some documentation.

Anyway, I am not sure which of those is better or worse.  There are advantages and disadvantages to each: the first option makes `SetWeights()` a little more complicated, but the second option adds another method to `Layer` that a user may need to overload.  Let me know what you think. :+1:

**Comment by rcurtin** [comment]
This will be fixed when #3231 is merged.  We went with the second approach (`CustomInitialize()`).
